server/players.py

```diff
@@ -89,6 +90,46 @@ def faction(self, value: Union[str, int, Faction]) -> None:
         else:
             self._faction = Faction.from_value(value)
 
+    @property
+    def vetoes(self) -> dict[int, int]:
+        return self._vetoes
+
+    @vetoes.setter
+    def vetoes(self, value: dict[int, int]) -> None:
+        if not isinstance(value, dict):
+            raise ValueError("Vetoes must be a dictionary")
+        if not all(isinstance(key, int) and isinstance(val, int) and val >= 0 for key, val in value.items()):
+            raise ValueError("Incorrect vetoes dictonary")
+        self._vetoes = value
+
+    async def update_vetoes(self, pools_vetodata: list[tuple[list[int], int, int]], current: dict = None) -> None:
```

Comment:
you could create a container (NamedTuple) for this `tuple[int], int, int` structure and avoid those long type hints and maybe call it `MapPoolVeto`

also, if you put tokens and ratings information in `MapPool` (in database they are in the same table, why are they separated here?) you won't even need this additional structure and you will be able to pass `MapPool` and use its properties

finally, this can also possibly remove `get_pools_veto_data` method since it just rearranges `MapPool`'s internals

Reply:
vetoes are not in map pool in database https://faforever.github.io/db/tables/map_pool.html
vetoes are here: https://faforever.github.io/db/tables/matchmaker_queue_map_pool.html
the same place where min and max rating, outside of map pool itself

Reply:
also i am not sure wdym removing get_pools_veto_data
it gives only necessary data while avoiding any code duplication and allowing to keep the code more clean
i am using it in multiple places so if we remove it code will become just worse

Reply:
Oh, i see

What i mean is that now `matchmaker_queue_map_pool` has additional 3 new fields and it may be viable to create corresponding data container (see, for example, `FeaturedMod`).
This way you preserve readable names and also make it easier to understand what's passed to other functions

For example, in my opinion

```python
async def update_vetoes(self, list[MatchmakerQueueMapPool], current: dict | None) -> None
```

is easier to follow than to find calls to `update_vetoes` and discover what's inside of `list[tuple[list[int], int, int]]`

also,
```python
map_pools: Iterable[MatchmakerQueueMapPool]
```
is easier to understand than
```python
map_pools: Iterable[tuple[MapPool, Optional[int], Optional[int], int, int, float]] = ()
```


where `MatchmakerQueueMapPool` is
```python
class MatchmakerQueueMapPool(NamedTuple):
    map_pool: MapPool
    min_rating: int | None
    max_rating: int | None
    veto_tokens_per_player: int
    max_tokens_per_map: int
    minimum_maps_after_veto: float
```

finally, this way long unpacking can be avoided and instead of 
```python
for pool, *_, veto_tokens_per_player, max_tokens_per_map, _ in queue.map_pools.values():
```
you can easily access named attributes
```python
for pool in queue.map_pools.values():
    pool.veto_tokens_per_player...

Reply:
well, yea, but this going into refactoring existing code zone rather than implementing new features so i am not sure should i do it here or not

Reply:
Using and changing to a NamedTuple should be fine and probably preferred as Gatsik says